lib_nn/src/cpp/filt2d/conv2d_utils.cpp

```diff
@@ -0,0 +1,240 @@
+
+#include "conv2d_utils.hpp"
+
+#include <algorithm>
+#include <cassert>
+#include <cmath>
+#include <cstdlib>
+#include <iostream>
+#include <vector>
+
+#include "xs3_vpu.h"
+
+using namespace nn;
+
+struct Conv2dChannelParams {
+  int16_t shift1;
+  int16_t scale;
+  int16_t offset_scale;
+  int16_t offset;
+  int16_t shift2;
+};
+
+static void compute_rshift_scale(int16_t &rshift, int16_t &scale,
+                                 const float &multiplier) {
+  rshift = 16;
+  float fscale = 0x7FFF;
+
+  if (multiplier != 0.0f) {
+    rshift = int16_t(-ceilf(log2f(multiplier))) + 1;
+    fscale = roundf(ldexpf(multiplier, 14 + rshift));
+  }
+
+  if (fscale == ldexpf(1, 15)) {
+    rshift -= 1;
+    fscale /= 2;
+  }
+
+  constexpr int SHIFT_ADJUSTMENT = 7;
+
+  // "We are using 16-bits instead of 8 so we need to adjust the shift"
+
+  rshift -= SHIFT_ADJUSTMENT;
+  scale = int16_t(fscale);
+}
+
+static Conv2dChannelParams computeConv2dChannelParams(
+    const float multiplier, const int32_t output_zero_point) {
+  // Conv2dChannelParams params;
+
+  int16_t rshift, scale;
+  compute_rshift_scale(rshift, scale, multiplier);
+
+  constexpr int SHIFT_ADJUSTMENT = 7;
```

Comment:
Do you think it would be better if these constants are moved to a header someplace else?